src/passes/Monomorphize.cpp

```diff
@@ -181,23 +397,159 @@ struct Monomorphize : public Pass {
+    // track this, map each old index to the new one.
+    std::unordered_map<Index, Index> mappedLocals;
+    auto newParamsMinusOld =
+      newFunc->getParams().size() - func->getParams().size();
+    for (Index i = 0; i < func->getNumLocals(); i++) {
+      if (func->isParam(i)) {
+        // Old params become new vars inside the function. Below we'll copy the
+        // proper values into these vars.
+        // TODO: We could avoid a var + copy when it is trivial (atm we rely on
+        //       optimizations to remove it).
+        auto local = Builder::addVar(newFunc.get(), func->getLocalType(i));
+        mappedLocals[i] = local;
+      } else {
+        // This is a var. The only thing to adjust here is that the parameters
+        // are changing.
+        mappedLocals[i] = i + newParamsMinusOld;
       }
     }
 
-    // Mark the chosen target in the map, so we don't do this work again: every
-    // pair of target and refinedParams is only considered once.
-    funcParamMap[{target, refinedParams}] = chosenTarget;
+    // Copy over local names to help debugging.
+    if (!func->localNames.empty()) {
+      newFunc->localNames.clear();
+      newFunc->localIndices.clear();
+      for (Index i = 0; i < func->getNumLocals(); i++) {
+        auto oldName = func->getLocalNameOrDefault(i);
+        if (oldName.isNull()) {
+          continue;
+        }
+
+        auto newIndex = mappedLocals[i];
+        auto newName = Names::getValidLocalName(*newFunc.get(), oldName);
+        newFunc->localNames[newIndex] = newName;
+        newFunc->localIndices[newName] = newIndex;
+      }
+    };
+
+    Builder builder(wasm);
+
+    // Surrounding the main body is the reverse-inlined content from the call
+    // context, like this:
+    //
+    //  (func $monomorphized
+    //    (..reverse-inlined parameter..)
+    //    (..old body..)
+    //  )
+    //
+    // For example, if a function that simply returns its input is called with a
+    // constant parameter, it will end up like this:
+    //
+    //  (func $monomorphized
+    //    (local $param i32)
+    //    (local.set $param (i32.const 42))  ;; reverse-inlined parameter
+    //    (local.get $param)                 ;; copied old body
+    //  )
+    //
+    // We need to add such an local.set in the prelude of the function for each
+    // operand in the context.
+    std::vector<Expression*> pre;
+    for (Index i = 0; i < context.operands.size(); i++) {
+      auto* operand = context.operands[i];
+
+      // Write the context operand (the reverse-inlined content) to the local
+      // we've allocated for this.
+      auto local = mappedLocals.at(i);
+      auto* value = ExpressionManipulator::copy(operand, wasm);
+      pre.push_back(builder.makeLocalSet(local, value));
+    }
+
+    // The main body of the function is simply copied from the original.
+    auto* newBody = ExpressionManipulator::copy(func->body, wasm);
```

Comment:
Haven't we already made a copy of the body when we copied the function? Why do we need another copy?

Reply:
Good catch, this was a leftover from an older approach. Fixed.